Runtime/Scripts/AudioStreamTrack.cs

```diff
@@ -35,13 +36,63 @@ public AudioClip Renderer
             get { return _streamRenderer?.clip; }
         }
 
+        internal class AudioBufferTracker
+        {
+            public const int NumOfFramesForBuffering = 5;
+            public long BufferPosition { get; set; }
+            public int SamplesPer10ms { get { return m_samplesPer10ms; } }
+
+            private readonly int m_sampleLength;
+            private readonly int m_samplesPer10ms;
+            private readonly int m_samplesForBuffering;
+            private long m_renderPos;
+            private int m_prevTimeSamples;
+
+            public AudioBufferTracker(int sampleLength)
+            {
+                m_sampleLength = sampleLength;
+                m_samplesPer10ms = m_sampleLength / 100;
+                m_samplesForBuffering = m_samplesPer10ms * NumOfFramesForBuffering;
+            }
+
+            public void Initialize(AudioSource source)
+            {
+                var timeSamples = source.timeSamples;
+                m_prevTimeSamples = timeSamples;
+                m_renderPos = timeSamples;
+                BufferPosition = timeSamples;
+            }
+
+            public int CheckNeedCorrection(AudioSource source)
+            {
+                if (source != null && m_prevTimeSamples != source.timeSamples)
+                {
+                    var timeSamples = source.timeSamples;
+                    m_renderPos += (timeSamples < m_prevTimeSamples ? m_sampleLength : 0) + timeSamples - m_prevTimeSamples;
+                    m_prevTimeSamples = timeSamples;
+
+                    if (m_renderPos >= BufferPosition)
+                    {
+                        return (int)(m_renderPos - BufferPosition) + m_samplesForBuffering;
+                    }
+                    else if (BufferPosition - m_renderPos <= m_samplesPer10ms)
+                    {
+                        return (int)(m_renderPos + m_samplesForBuffering - BufferPosition);
+                    }
+                }
+
+                return 0;
+            }
+        }
+
 
         internal class AudioStreamRenderer : IDisposable
         {
-            private const int SamplesPerSec = 100;
-
             private AudioClip m_clip;
-            private readonly BlockingCollection<float[]> m_recvBufs = new BlockingCollection<float[]>(SamplesPerSec);
+            private bool m_bufferReady = false;
+            private readonly Queue<float[]> m_recvBufs = new Queue<float[]>();
+            private readonly AudioBufferTracker m_bufInfo;
+            private AudioSource m_attachedSource;
```

Comment:
`m_attachedSource` instance should be set from `AudioStreamRenderer` constructor.
There is a problem that finding the instance in `SetData` method.

Reply:
> There is a problem that finding the instance in `SetData` method.
I'd like to ask what the problem is? because it's a intentional lazy initialization to wait until that m_clip attached by AudioSource.

Reply:
`GameObject.FindObjectsOfType` method is used in this line. 
https://github.com/Unity-Technologies/com.unity.webrtc/pull/565/files#diff-3eff72b49e750b2ef3b0f9311a838c24f98ddcfbd5482d1db99df5ce84d7a521R125

I feel setting `AudioSource` from constractor is better than using the Find* method.

Reply:
I'd like to ask you why is it better to use Find* method in constructor? I don't have much experience with Unity, so if you explain in more detail, I can understand your concerns.

Reply:
OK, please let me explain more detail.
My suggestion is making constructor having `AudioSource` argument, and removing `FindAttachedAudioSource` method.

```
public AudioStreamRenderer(AudioSource source)
```

Generally, `GameObject.Find*` has worse performance comparing to direct referencing. There are some projects restricting using  `GameObject.Find*` to avoid performance issue. Additionally, using  `GameObject.Find*` decrease readability and testability.

These links are just a few examples.
[https://dev.to/andytq/unity-comparison-of-performance-cost-between-gameobject-find-and-direct-referencing-pb](https://dev.to/andytq/unity-comparison-of-performance-cost-between-gameobject-find-and-direct-referencing-pb)
[https://akbiggs.silvrback.com/please-stop-using-gameobject-find](https://akbiggs.silvrback.com/please-stop-using-gameobject-find)
[https://gamedevunboxed.com/why-using-unitys-find-functions-is-a-mistake/](https://gamedevunboxed.com/why-using-unitys-find-functions-is-a-mistake/)

If you are OK, Could you allow me to refactor your code?

Reply:
As you said, to avoid using `FindObjectsOfType`, passing AudioSource as argument was considered too. but it makes breaking changes to it's usage, so I had to use `FindObjectsOfType` for current interface compatibility.
Anyway, please feel free to make any modification whatever it is, welcome better solution always. 😃